File: src/main/java/com/ptah/common/converter/StringListConverter.java
package com.ptah.common.converter;

import com.ptah.common.util.StringUtils;
import jakarta.persistence.AttributeConverter;
import jakarta.persistence.Converter;

import java.util.Collections;
import java.util.Optional;
import java.util.Set;

@Converter
public class StringListConverter implements AttributeConverter<Set<String>, String> {

    private static final String SPLIT_CHAR = ",";


    @Override
    public String convertToDatabaseColumn(Set<String> attribute) {
        return Optional.ofNullable(attribute).map(value -> String.join(SPLIT_CHAR, value)).orElse(StringUtils.EMPTY);
    }

    @Override
    public Set<String> convertToEntityAttribute(String dbData) {
        return Optional.ofNullable(dbData).map(data -> Set.of(data.split(SPLIT_CHAR))).orElseGet(Collections::emptySet);
    }


}

File: src/main/kotlin/com/ptah/common/converter/StringListConverter.kt
package com.ptah.common.converter

import jakarta.persistence.AttributeConverter
import jakarta.persistence.Converter
import org.apache.commons.lang3.StringUtils
import java.util.*

@Converter
class StringListConverter : AttributeConverter<Set<String>, String> {
    override fun convertToDatabaseColumn(attribute: Set<String>): String = attribute.joinToString(SPLIT_CHAR)
    override fun convertToEntityAttribute(dbData: String): Set<String> = dbData.split(SPLIT_CHAR).toSet()

    companion object {
        private const val SPLIT_CHAR = ","
    }
}
